lib/scholar/cluster/k_means.ex

```diff
@@ -0,0 +1,325 @@
+         broadcast_x, centroids, nearest_centroids}
+      end
+
+    {_inertia_for_centroids, min_inertia} =
+      calculate_inertia(x, final_centroids, num_clusters, num_runs)
+
+    final_inertia = (min_inertia * weights) |> Nx.sum(axes: [1])
+    best_run = Nx.argmin(final_inertia)
+
+    %__MODULE__{
+      num_iterations: i,
+      clusters: final_centroids[best_run],
+      labels: nearest_centroids[best_run],
+      inertia: final_inertia[best_run]
+    }
+  end
+
+  defnp initialize_centroids(x, opts) do
+    num_clusters = opts[:num_clusters]
+    {num_samples, _num_features} = Nx.shape(x)
+    num_runs = opts[:num_runs]
+
+    case opts[:init] do
+      :random ->
+        Nx.iota({num_runs, num_samples}, axis: 1)
+        |> Nx.shuffle(axis: 1)
+        |> Nx.slice_along_axis(0, num_clusters, axis: 1)
+        |> then(&Nx.take(x, &1))
+
+      :k_means_plus_plus ->
+        k_means_plus_plus(x, num_clusters, num_runs)
+    end
+  end
+
+  defnp calculate_inertia(x, centroids, num_clusters, num_runs) do
+    {num_samples, num_features} = Nx.shape(x)
+
+    modified_centroids =
+      centroids
+      |> Nx.reshape({num_runs, num_clusters, 1, num_features})
+      |> Nx.broadcast({num_runs, num_clusters, num_samples, num_features})
+      |> Nx.reshape({num_runs, num_clusters * num_samples, num_features})
+
+    inertia_for_centroids =
+      Scholar.Cluster.Utils.squared_euclidean(
+        Nx.tile(x, [num_runs, num_clusters, 1]),
+        modified_centroids
+      )
+      |> Nx.reshape({num_runs, num_clusters, num_samples})
+
+    {inertia_for_centroids, Nx.reduce_min(inertia_for_centroids, axes: [1])}
+  end
+
+  defnp generate_idx(mod_weights, num_runs) do
+    rand = Nx.random_uniform({num_runs}, 0, 1, type: {:f, 32})
+    val = (rand * Nx.sum(mod_weights, axes: [1])) |> Nx.new_axis(-1)
+    cumulative_weights = Nx.cumulative_sum(mod_weights, axis: 1)
+    (val <= cumulative_weights) |> Nx.argmax(tie_break: :low, axis: 1)
+  end
+
+  defnp find_new_centroid(weights, x, centroid_mask, num_runs) do
+    mod_weights = weights * centroid_mask
+    idx = generate_idx(mod_weights, num_runs)
+    {Nx.take(x, idx), idx}
+  end
+
+  defnp k_means_plus_plus(x, num_clusters, num_runs) do
+    inf = Nx.Constants.infinity()
+    {num_samples, num_features} = Nx.shape(x)
+    centroids = Nx.broadcast(inf, {num_runs, num_clusters, num_features})
+    inertia = Nx.broadcast(0.0, {num_runs, num_samples})
+    centroid_mask = Nx.broadcast(1, {num_runs, num_samples})
+
+    first_centroid_idx = Nx.random_uniform({num_runs}, 0, num_samples - 1, type: {:u, 32})
+
+    first_centroid = Nx.take(x, first_centroid_idx) |> Nx.reshape({:auto})
+
+    indices_centroids =
+      Nx.concatenate([
+        Nx.new_axis(Nx.reshape(Nx.iota({num_runs, num_features}, axis: 0), {:auto}), 0),
+        Nx.new_axis(Nx.broadcast(0, {num_runs * num_features}), 0),
+        Nx.new_axis(Nx.tile(Nx.iota({num_features}), [num_runs]), 0)
+      ])
```

Comment:
Can't this be Nx.stack instead?

Also, if we can avoid the transpose it would be better